glib/src/futures/mod.rs

```diff
@@ -0,0 +1,164 @@
+use futures_task::Poll;
+use futures_util::future::Either;
+use futures_util::future::Select;
+use futures_util::pin_mut;
+
+use crate::FutureWithTimeoutError;
+use crate::Priority;
+use std::future::{Future, IntoFuture};
+use std::{pin::Pin, time::Duration};
+
+#[derive(Default)]
+pub enum SchedulingPrecision {
+    #[default]
+    Millisecond,
+    Second,
+}
+
+pub struct Sleep {
+    duration: Duration,
+    priority: Priority,
+    precision: SchedulingPrecision,
+}
+
+impl IntoFuture for Sleep {
+    type Output = ();
+
+    type IntoFuture = Pin<Box<dyn Future<Output = ()> + Send + 'static>>;
+
+    fn into_future(self) -> Self::IntoFuture {
+        use SchedulingPrecision::*;
+        match self.precision {
+            Millisecond => crate::timeout_future_with_priority(self.priority, self.duration),
+            Second => crate::timeout_future_seconds_with_priority(
+                self.priority,
+                self.duration.as_secs_f32().round() as u32,
+            ),
+        }
+    }
+}
+
+impl Sleep {
+    pub fn priority(mut self, priority: Priority) -> Self {
+        self.priority = priority;
+        self
+    }
+    pub fn precision(mut self, precision: SchedulingPrecision) -> Self {
```

Comment:
Maybe the precision should be defined based on whether the `Duration` has fractional seconds or not? Or is that too implicit?

Otherwise maybe having a `sleep(Duration)` and `sleep_seconds(u32)` would be simpler API.

Reply:
We can't make it implicit based on the choosen `Duration`, because the user could want to actually wait `2000` milliseconds precisely, and not `2` seconds roughly.

What about having the methods `precision_seconds`, `precision_milliseconds`, so that we can remove the public enum `SchedulingPrecision`?

Reply:
Good point. Yes that would also work

Reply:
Thinking about it... In gtk adding a new enum for a specific function seems to be pretty normal. I think it fits adding an enum here as well